src/components/view-pane/index.tsx

```diff
@@ -1,32 +1,31 @@
-import {getTopSpecQueryItem, SpecQueryGroup} from 'compassql/build/src/model';
+import {SpecQueryGroup} from 'compassql/build/src/model';
 import {Query} from 'compassql/build/src/query/query';
 import * as React from 'react';
 import * as CSSModules from 'react-css-modules';
 import {connect} from 'react-redux';
-import {Data} from 'vega-lite/build/src/data';
-import {OneOfFilter, RangeFilter} from 'vega-lite/build/src/filter';
+import {FacetedCompositeUnitSpec} from 'vega-lite/build/src/spec';
 import {ActionHandler, createDispatchHandler} from '../../actions/redux-action';
 import {ShelfAction} from '../../actions/shelf';
 import {State} from '../../models';
 import {Bookmark} from '../../models/bookmark';
 import {extractPlotObjects, PlotObject} from '../../models/plot';
-import {getTransforms, hasWildcards} from '../../models/shelf/spec';
-import {getBookmark, getData, getFilters, getMainResult, getQuery} from '../../selectors';
+import {hasWildcards} from '../../models/shelf/spec';
+import {getBookmark, getMainResult, getQuery, getSpec} from '../../selectors';
 import {Plot} from '../plot';
 import {PlotList} from '../plot-list';
 import * as styles from './view-pane.scss';
 
+
 export interface ViewPaneProps extends ActionHandler<ShelfAction> {
-  data: Data;
   query: Query;
-  filters: Array<RangeFilter | OneOfFilter>;
   mainResult: SpecQueryGroup<PlotObject>;
   bookmark: Bookmark;
+  spec: FacetedCompositeUnitSpec;
 }
 
 class ViewPaneBase extends React.PureComponent<ViewPaneProps, {}> {
   public render() {
-    const {bookmark, data, handleAction, filters, query, mainResult} = this.props;
+    const {bookmark, handleAction, query, mainResult, spec} = this.props;
```

Comment:
`query` and `mainResult` should no longer be a prop of this component. 

We should instead have `mainSpec` and `mainPlots` and at most one of them can have value at a particular time.

The selector for `mainSpec` and `mainPlots` should depend on `query` and the full `mainResult`  -- not just `result.main.modelGroup`.  Thus, we can read `result.main.isLoading`.

If `result.main.isLoading` is`true`, both `mainSpec` and `mainPlots` should be `undefined`.